aesd-char-driver/main.c

```diff
@@ -85,159 +96,340 @@ ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+        kfree(placeholder);
 
-                if(cb_fifo->full)
-                {
-                    kfree(cb_fifo->entry[cb_fifo->out_offs].buffptr);
-                }
-                aesd_circular_buffer_add_entry(cb_fifo, &newEntry);
-
-                PDEBUG("%zu bytes at %p in cbfifo",newEntry.size,newEntry.buffptr);
+        if (nlFlag)
+        {
+            // Add new entry to CB
+            newEntry.buffptr = newLimboString;
+            newEntry.size = aesd_device.limboLength + entrySize;
 
-                //Get device out of packet limbo
-                aesd_device.inLimbo = false;
-                aesd_device.limboLength = 0;
-                kfree(aesd_device.limboString);
-            }
-            else
+            if (cb_fifo->full)
             {
-                //Update limbo string
-                kfree(aesd_device.limboString);
-                aesd_device.limboString = newLimboString;
-                aesd_device.limboLength += entrySize;  
+                kfree(cb_fifo->entry[cb_fifo->out_offs].buffptr);
             }
+            aesd_circular_buffer_add_entry(cb_fifo, &newEntry);
+
+            PDEBUG("%zu bytes at %p in cbfifo", newEntry.size, newEntry.buffptr);
+
+            // Get device out of packet limbo
+            aesd_device.inLimbo = false;
+            aesd_device.limboLength = 0;
+            kfree(aesd_device.limboString);
+        }
+        else
+        {
+            // Update limbo string
+            kfree(aesd_device.limboString);
+            aesd_device.limboString = newLimboString;
+            aesd_device.limboLength += entrySize;
+        }
     }
     else
     {
-        if(nlFlag)
+        if (nlFlag)
         {
-            //Clean packet entry scenario
+            // Clean packet entry scenario
             newEntry.buffptr = placeholder;
             newEntry.size = entrySize;
 
-            if(cb_fifo->full)
+            if (cb_fifo->full)
             {
                 kfree(cb_fifo->entry[cb_fifo->out_offs].buffptr);
             }
             aesd_circular_buffer_add_entry(cb_fifo, &newEntry);
         }
         else
         {
-            //Enter limbo
+            // Enter limbo
             aesd_device.inLimbo = true;
             aesd_device.limboString = placeholder;
             aesd_device.limboLength = entrySize;
         }
     }
-    retval=entrySize;
-    
-    out:
-        mutex_unlock(&(aesd_device.lock));
-        return retval;
+    retval = entrySize;
+
+out:
+    mutex_unlock(&(aesd_device.lock));
+    return retval;
+}
+
+loff_t aesd_llseek(struct file *filp, loff_t off, int whence)
+{
+    loff_t newpos;
+    struct aesd_buffer_entry *entryPtr = NULL;
+    struct aesd_circular_buffer *devBuffPtr = &aesd_device.dev_cb_fifo;
+    size_t entryPtrInd;
+
+    uint64_t devBuffSize;
+
+    switch (whence)
+    {
+    case 0: /* SEEK_SET */
+        entryPtr = aesd_circular_buffer_find_entry_offset_for_fpos(
```

Comment:
@ECleverito what's your locking strategy here?  I think you are missing some locking to protect `devBuffPtr`.